keps/sig-node/3857-rro-mounts/README.md

```diff
@@ -0,0 +1,1080 @@
++	RecursivelyReadOnlyIfPossible RecursivelyReadOnlyMode = "IfPossible"
++	// RecursivelyReadOnlyEnabled enables recursively-readonly mode, or raise an error.
++	RecursivelyReadOnlyEnabled RecursivelyReadOnlyMode = "Enabled"
++)
++
+ // VolumeDevice describes a mapping of a raw block device within a container.
+ type VolumeDevice struct {
+ 	// name must match the name of a persistentVolumeClaim in the pod
+@@ -2505,6 +2524,10 @@ type ContainerStatus struct {
+ 	// +featureGate=InPlacePodVerticalScaling
+ 	// +optional
+ 	Resources *ResourceRequirements
++	// Status of volume mounts.
++	// +listType=atomic
++	// +optional
++	VolumeMounts []VolumeMountStatus
+ }
+ 
+ // PodPhase is a label for the condition of a pod at the current time.
+@@ -2584,6 +2607,24 @@ const (
+ 	PodResizeStatusInfeasible PodResizeStatus = "Infeasible"
+ )
+ 
++// VolumeMountStatus shows status of volume mounts.
++type VolumeMountStatus struct {
++	// Corresponds to the name of the original VolumeMount.
++	Name string
++	// Corresponds to the mount path of the original VolumeMount.
++	MountPath string
++	// ReadOnly must be set to true or false.
++	// A nil value in the original VolumeMount must be translated to true or false,
++	// depending on the mount result.
++	ReadOnly bool
++	// RecursivelyReadOnly must be set to Disabled, Enabled, or nil (for non-readonly mounts).
++	// An IfPossible value in the original VolumeMount must be translated to Disabled or Enabled,
++	// depending on the mount result.
++	// +featureGate=RecursivelyReadOnlyMounts
++	// +optional
++	RecursivelyReadOnly *RecursivelyReadOnlyMode
++}
++
+ // RestartPolicy describes how the container should be restarted.
+ // Only one of the following restart policies may be specified.
+ // If none of the following policies is specified, the default one
+@@ -4446,6 +4487,17 @@ type NodeDaemonEndpoints struct {
+ 	KubeletEndpoint DaemonEndpoint
+ }
+ 
++// RuntimeClassInfo is a set of runtime class information.
++type RuntimeClassInfo struct {
++	// Runtime class name.
++	// Empty for the default runtime class.
++	// +optional
++	Name string
++	// RecursivelyReadOnlyMounts is set to true if the runtime class supports RecursivelyReadOnlyMounts.
++	// +optional
++	RecursivelyReadOnlyMounts *bool
++}
++
+ // NodeSystemInfo is a set of ids/uuids to uniquely identify the node.
+ type NodeSystemInfo struct {
+ 	// MachineID reported by the node. For unique machine identification
+@@ -4472,6 +4524,9 @@ type NodeSystemInfo struct {
+ 	OperatingSystem string
+ 	// The Architecture reported by the node
+ 	Architecture string
++	// The available runtime classes.
++	// +optional
++	RuntimeClasses []RuntimeClassInfo
+ }
+ 
+ // NodeConfigStatus describes the status of the config assigned by Node.Spec.ConfigSource.
+```
+
+### CRI API
+
+Add `recursively_read_only=(DISABLED|BEST_EFFORT|ENABLED)` (enum) to the [`Mount`](https://github.com/kubernetes/cri-api/blob/v0.26.1/pkg/apis/runtime/v1/api.proto#L212-L224) message.
```

Comment:
does BEST_EFFORT need to be exposed if the runtime reports what it supports?

Reply:
No.
Updated to use a plain boolean in the CRI API.